04-materials/03-debugging.md

```diff
@@ -1,111 +1,431 @@
 
-### Types of Errors
-- Build-time errors vs runtime errors
-- Python backend errors vs TypeScript frontend errors
+- **Identifying the source of the error**:
+  When JavaScript throws an error, it gives you a stack trace, a trail showing the path the code took before it crashed. It may seem overwhelming at first, but starting from the top and looking for recognizable file names can help. While the stack trace might show you multiple function calls usually only one or two of those are in _your_ code. The others are likely in other libraries like JupyterLab itself. Finding your line in the code, you can find out where things have gone wrong.
 
-### Reading Stack Traces
-- Minified vs source-mapped traces
-- JavaScript stack traces
-- Identifying the source of the error
+- **Minified vs source-mapped traces**:
+  Also important, if you see names like `bundle.js:4323`, that's minified code. In development mode with {term}`source maps <Source maps>` enabled, you'll see actual files names like `Panel.tsx:53`, a much more helpful identifier. Extensions created with the `extension-template` should have source maps enabled already.
 
-## 2. Terminal
+## 2️⃣ Terminal
+
+### 📋 Terminal Logs
 
-### Terminal Logs
 - Where to find JupyterLab server logs
+  In the terminal where you run `jupyter lab`, you'll find a great deal of information.
+
+During build time, webpack will tell you about compilation errors, missing dependencies, or TypeScript type errors. And while the messages can be verbose, you can scroll up to the first error and work through any subsequent errors.
+
+## 3️⃣ Browser Developer Tools
+
+This is where the real magic happens! Head to your browser and open the developer tools. Let's take a look at some of the most useful tabs.
 
-## 3. Browser Developer Tools
+:::{tip} 🔧 How to open DevTools
+
+- **Windows/Linux**: Press `F12` or `Ctrl+Shift+I`
+- **Mac**: Press `Cmd+Option+I`
+  :::
+
+### 💬 Console Tab
 
-### Console Tab
 - Viewing runtime errors and warnings
-- Using `console.log()`, `console.error()`, and `console.warn()`
-    - Filtering console messages
-- Preserving logs across reloads
-- Structured logging for better debugging
-- JupyterLab's internal logging <!-- TODO: brief, make reference to section 02 -->
-
-### Network Tab
-- Checking extension resource loading
-- Monitoring API requests to JupyterLab backend
-- Inspecting request/response payloads
-- Debugging failed network requests
-
-### Sources Tab with Source Maps
-- Source maps are enabled by default if you're working from the extension-template
-- Setting breakpoints
-- Stepping through code execution
-
-### Elements/Inspector Tab
+
+   Sometimes, the best tool is {term}`print debugging <Print debugging>`, which is like a print statement for the browser console, using a simple `console.log()`. Walking through your changes and adding a logging statement to verify/inspect data, or ensure your code reaches certain logic can be quick and effective.
```

Comment:
I'm not sure why this term reference isn't working. It looks right to me!

Reply:
Thanks for catching this Matt! It turns out I added these terms after the closing `:::` of the glossary block 😅